File: src/XCocoaUtilsPublic/UIKit/src/UIViewController+XCUP.m
//
//  UIViewController+Utils.m
//  MKTest
//
//  Created by Deheng.Xu on 12-11-5.
//  Copyright (c) 2012年 Deheng.Xu. All rights reserved.
//

#import "UIViewController+XCUP.h"
#import "NSObject+XCUP.h"
#import "XCUPMacros.h"

#define ASSERT_LOADING  0

#if TARGET_OS_IPHONE

#import <UIKit/UIKit.h>

@implementation UIViewController (XCUP)

//+ (NSString *)xibFileNameDefaultSuffix
//{
//    return ([[[[UIDevice currentDevice] model] lowercaseString] rangeOfString:@"ipad"].length > 0) ? @"iPad" : @"iPhone";
//}

//+ (instancetype)viewControllerWithNibName:(NSString *)nibName bundle:(NSBundle *)nibBundleOrNil
//{
//    id rtn = nil;
//    NSString *suffix = [self xibFileNameDefaultSuffix];
//    
//#if ASSERT_LOADING
//        NSAssert(([[NSFileManager defaultManager] fileExistsAtPath:[[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_%@", nibName, suffix] ofType:@"nib"]] == YES), @"Nib file '%@' does not exist!", nibName);
//#endif
//    if (![[NSFileManager defaultManager] fileExistsAtPath:[[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_%@", nibName, suffix] ofType:@"nib"]]) {
//        rtn = [[self alloc] initWithNibName:nibName bundle:nibBundleOrNil];
//    }else {
//        rtn = [[self alloc] initWithNibName:[NSString stringWithFormat:@"%@_%@", nibName, suffix] bundle:nibBundleOrNil];
//    }
//    return XAutorelease(rtn);
//}

//- (BOOL)isSupportInteractivePopGestureRecognizer
//{
//    return ([self isKindOfClass:[UINavigationController class]] && [self respondsToSelector:@selector(interactivePopGestureRecognizer)]);
//}

//- (UINavigationController *)xcup_navigationController {
//	if (self.navigationController) {
//		return self.navigationController;
//	}else {
//		UINavigationController* nav = [[UINavigationController alloc] initWithRootViewController:self];
//		return nav;
//	}
//}

//- (instancetype)xcup_present:(UIViewController *)presentedViewController animated:(BOOL)animated orNeedNavigation:(BOOL)needed completion:(void (^)(void))completion {
//	if (self.navigationController) {
//		if (needed) {
//			[self.navigationController pushViewController:presentedViewController animated:animated];
//		}else {
//			[self presentViewController:presentedViewController animated:animated completion:completion];
//		}
//	}else {
//		if (needed) {
//            if ([self isKindOfClass:UINavigationController.class]) {
//                [(UINavigationController*) self pushViewController:presentedViewController animated:animated];
//            }else {
//                UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:presentedViewController];
//                [self presentViewController:nav animated:animated completion:completion];
//            }
//		}else {
//			[self presentViewController:presentedViewController animated:animated completion:completion];
//		}
//	}
//	return self;
//}

//- (instancetype)xcup_present:(UIViewController *)presentedViewController animated:(BOOL)animated modalStyle:(UIModalPresentationStyle)presentationStyle transitionStyle:(UIModalTransitionStyle)transitionStyle orNeedNavigation:(BOOL)needed completion:(void (^)(void))completion {
//	if (self.navigationController) {
//		if (needed) {
//			[self.navigationController pushViewController:presentedViewController animated:animated];
//		}else {
//			presentedViewController.modalPresentationStyle = presentationStyle;
//			presentedViewController.modalTransitionStyle = transitionStyle;
//			[self presentViewController:presentedViewController animated:animated completion:completion];
//		}
//	}else {
//		if (needed) {
//			if ([self isKindOfClass:UINavigationController.class]) {
//				[(UINavigationController*) self pushViewController:presentedViewController animated:animated];
//			}else {
//				UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:presentedViewController];
//				nav.modalPresentationStyle = presentationStyle;
//				nav.modalTransitionStyle = transitionStyle;
//				[self presentViewController:nav animated:animated completion:completion];
//			}
//		}else {
//			[self presentViewController:presentedViewController animated:animated completion:completion];
//		}
//	}
//	return self;
//}

//- (instancetype)xcup_present:(UIViewController *)presentedViewController animated:(BOOL)animated completion:(void (^)(void))completion {
//	if (presentedViewController.navigationController) {
//		[self presentViewController:presentedViewController.navigationController animated:animated completion:completion];
//	}else {
//		[self presentViewController:presentedViewController animated:animated completion:completion];
//	}
//	return self;
//}

//- (instancetype)xcup_dismissViewController:(BOOL)animated completion:(void (^)(void))completion {
//	if (self.navigationController) {
//		[self.navigationController dismissViewControllerAnimated:animated completion:completion];
//	}else {
//		[self dismissViewControllerAnimated:animated completion:completion];
//	}
//	return self;
//}

@end
#endif

File: src/XCocoaUtilsPublic/Swift/src/UIViewController+XCUP.swift
//
//  UIViewController+XCUP.swift
//  XCocoaUtilsPublic
//
//  Created by Deheng Xu on 2021/12/15.
//

#if os(iOS)
import UIKit

@objc public extension UIViewController {

	static func xibFileNameDefaultSuffix() -> String {
		return UIDevice.current.model.lowercased().hasSuffix("ipad") ? "iPad" : "iPhone"
	}

	static func viewController(nibName: String, bundle bundleOrNil: Bundle) -> UIViewController {
		if let path = Bundle.main.path(forResource: "\(nibName)_\(xibFileNameDefaultSuffix())", ofType: "nib"), !FileManager.default.fileExists(atPath: path) {
			return self.init(nibName: nibName, bundle: bundleOrNil)
		} else {
			return self.init(nibName: "\(nibName)_\(xibFileNameDefaultSuffix())", bundle: bundleOrNil)
		}
	}

	func withNavgationAsPossible() -> UINavigationController {
		if let nav = self.navigationController {
			return nav
		} else {
			let nav = UINavigationController(rootViewController: self)
			return nav
		}
	}

	func isSupportInteractivePopGestureRecognizer() -> Bool {
		if let _ = self as? UINavigationController, #available(iOS 7.0, *) {
			return true
		}
		return false
	}


	func xcup_navigationController() -> UINavigationController {
		if let nav = self.navigationController {
			return nav
		} else {
			let nav = UINavigationController(rootViewController: self)
			return nav
		}
	}

	func xcup_present(_ viewController: UIViewController, animated: Bool = true, presentationStyle: UIModalPresentationStyle = .fullScreen, transitionStyle:UIModalTransitionStyle = .coverVertical, orNeedNavigation: Bool = false, completion: (() -> Void)? = nil) {
		var targetVc: UIViewController = viewController
		if let nav = self.navigationController {
			targetVc = nav
		} else if orNeedNavigation {
			targetVc = UINavigationController(rootViewController: viewController)
		}
		targetVc.modalTransitionStyle = transitionStyle
		targetVc.modalPresentationStyle = presentationStyle
		self.present(targetVc, animated: animated, completion: completion)
	}

	func xcup_dismiss(animated: Bool = true, completion: (() -> Void)? = nil) {
		if let nav = self.navigationController {
			nav.dismiss(animated: animated, completion: completion)
		} else {
			self.dismiss(animated: animated, completion: completion)
		}
	}

	func setNavigationBarColor(_ color: UIColor, for barMetrics: UIBarMetrics = .default) {
		self.navigationController?.navigationBar.setBackgroundImage(UIImage(color: color, size: self.navigationController?.navigationBar.frame.size ?? .zero), for: barMetrics)
	}
}

#endif
